README.md

# TRES <img src="man/figures/TRES_logo.png" align="right" width="200"/>

<!-- badges: start -->
[![cran webpage](https://img.shields.io/badge/cran-1.1.5-blue)](https://CRAN.R-project.org/package=TRES)
<!-- badges: end -->

The package **TRES** implements the least squares and envelope estimation under the framework of tensor regression models. The general model-free envelope models can also be flexibly handled by the package via three types of envelope estimation algorithms: 
- Full Grassmannian (FG) algorithm.
- 1D algorithm.
- Envelope coordinate descent (ECD) algorithm
- Partial least squares (PLS) type algorithm.

## Installation

You can install the released version of TRES from [CRAN](https://CRAN.R-project.org) with:

``` r
# Install the latest released version from CRAN
install.packages("TRES")

# Or the development version from GitHub:
remotes::install_github("leozeng15/TRES")
```

## Example 1: Tensor response regression analysis

This is a basic example providing you a guidance on how to use the primary function `TRR.fit` and several **S**3 methods in Tensor Response Regression (TRR) model. The ordinary least square method and 1D envelope method are implemented. See Li and Zhang (2017) for more background.

``` r
library(TRES)
## Load data "bat"
data("bat")
x <- bat$x
y <- bat$y

## Fitting with OLS and 1D envelope method.
fit_ols <- TRR.fit(x, y, method="standard")
fit_1D <- TRR.fit(x, y, u = c(14,14), method="1D") # pass envelope rank (14,14)

## Print cofficient
coef(fit_1D)

## Print the summary
summary(fit_1D)

## Extract the mean squared error, p-value and standard error from summary
summary(fit_1D)$mse
summary(fit_1D)$p_val
summary(fit_1D)$se

## Make the prediction on the original dataset
predict(fit_1D, x)

## Draw the plots of two-way coefficient tensor (i.e., matrix) and p-value tensor.
plot(fit_ols)
plot(fit_1D)
```

The coefficients plots from OLS and 1D methods are aligned in the first row below, and the p-value plots from the two methods are aligned in the second row below.

<div>
<img src="man/figures/resp_2D_std.png" width="300"/> <img src="man/figures/resp_2D_1D.png" width="300"/> <br>
 <img src="man/figures/p_value_std.png" width="300"/><img src="man/figures/p_value_1D.png" width="300"/>
</div>

## Example 2: Model-free envelope estimation
This example shows how to use the function `MenU_sim` to simulate the matrices `M` and `U` with envelope structure, and how to use different core functions to implement different envelope estimation algorithms. See Cook and Zhang (2016) for more details.
``` r
## Generate matrices M and U
p <- 50
u <- 5
n <- 200
data <- MenvU_sim(p, u, jitter = 1e-5, wishart = TRUE, n = n)
Gamma <- data$Gamma
M <- data$M
U <- data$U

## Use different envelope algorithms
G <- vector("list", 6)
G[[1]] <- simplsMU(M, U, u)
G[[2]] <- ECD(M, U, u)
G[[3]] <- manifold1D(M, U, u)
G[[4]] <- OptM1D(M, U, u)
G[[5]] <- manifoldFG(M, U, u)
G[[6]] <- OptMFG(M, U, u)
```

## References

*1D algorithm*: Cook, R.D. and Zhang, X., 2016. Algorithms for envelope estimation. *Journal of Computational and Graphical Statistics*, 25(1), pp.284-300.

*TRR*: Li, L. and Zhang, X., 2017. Parsimonious tensor response regression. *Journal of the American Statistical Association*, 112(519), pp.1131-1146.

*TPR*: Zhang, X. and Li, L., 2017. Tensor envelope partial least-squares regression. *Technometrics*, 59(4), pp.426-436.


*ECD algorithm*: Cook, R.D. and Zhang, X., 2018. Fast envelope algorithms. *Statistica Sinica*, 28(3), pp.1179-1197.

## Journal of Statistical Software paper
Zeng J., Wang W., Zhang X. (2021) TRES: An R Package for Tensor Regression and Envelope Algorithms. *Journal of Statistical Software*, 99(12), 1-31. doi:10.18637/jss.v099.i12



NEWS.md
# TRES 1.1.5

## Major changes
- Revise the help documentation for the function `OptStiefelGBB.R`.
- The Journal of Statistical Software paper accompanying this package has been published.

# TRES 1.1.4

## Major changes
- We change the title from ``Tensor Regression with Envelope Structure and Three Generic Envelope Estimation Approaches'' to ``Tensor Regression with Envelope Structure''.

## Minor change
- Revert `res <- kroncov(En)` to `res <- try(kroncov(En))` in function `TRR.fit`, which is forgotten to be reverted in last version.


# TRES 1.1.3

## Major changes
- Replace the `base::norm(x, 'F')` function with `sqrt(sum(x^2))` for some matrix `x`.
- In the `Data` segment of the printout of the `tensor` object, instead of just the first few elements, the summarized results of all elements are printed.

## Minor changes
- Add more output information returned from `OptManiMulitBallGBB()` to the output of `ballGBB1D()`.

---

# TRES 1.1.2

## Major changes

- Renaming: `TRR_sim -> TRRsim; TPR_sim -> TPRsim; TensEnv_dim -> TRRdim; ballGBB1D_bic -> oneD_bic; TensPLS_cv2d3d -> TPRdim; OptimballGBB1D -> OptM1D; EnvMU -> simplsMU`
- Deprecate:
- `plot.Tenv`:
    - Correct the direction of y axis. The coordinates of x and y increase from left to right, top to bottom; 
    - Add arguments `xlab = ""`, `ylab = ""`, `axes = TRUE`, `ask = TRUE`, remove arguments `xticks`, `yticks`. `axes` is a logical value specifying whether the axes should be drawn. If `ask = TRUE`, user is prompted before the second plot is shown (if exists).
- `ECD, simplsMU, manifold1D, manifoldFG, OptM1D, OptMFG`: Change optional arguments like `maxiter`, `tol`, to three dots `...`
- `fun1D, get_ini1D, ballGBB1D, OptManiMulitBallGBB`: Collected into file "1Dfunction.R". Help documentation are no longer supported.
- `MenvU_sim`: An option `wishart = FALSE` is provided to return the population matrices `M` and `U`. Other argument like `jitter` is also provided which adds an scaler matrix to `M` to ensures it positive-definiteness.
- `PMSE`: Calculates the prediction and mean squared error for both TRR and TPR models.
- `kroncov`: Add convergence criterion `tol` and the maximal iteration `maxiter`.


## Minor changes
- `OptStiefelGBB`: Hide `out` from the output.
- `manifoldFG`: Set default value for `Gamma_init`.
- `ttt.R`: Change arguments `X,Y` to `x,y`.
- `oneD_bic`: Add the estimated envelope basis `Gamma` to the output.
- `TRRdim`: Output the mean squared error using the selected envelope basis.
- `TRRdim, oneD_bic`: Argument `multiD` is changed to `C` to comply with the paper Zhang X, Mai Q (2018). "Model-Free Envelope Dimension Selection." Electronic Journal of Statistics.

### New functions
- `OptMFG`: New FG optimization function encapsulating the core function `OptStiefelGBB`.
- `show`: overloads `show` in package `rTensor`. With the overloaded `show`, only  the first 6 elements of tensor is printed out.

## Bugs
- In documentation of `subspace()`, the formula of subspace distance should be ||P_{A} - P_{B}||_F/√{2d}.
- Fix bat dataset where `x` was not binary.

---

# TRES 1.1.1

## Major changes
- Lists or environments data structures are also acceptable as a whole in `TRR.fit()` and `TPR.fit()`. They can be passed to argument `x`.
- Since the variance-covariance matrix is not available for `Tenv` class object, we remove S3 methods `vcov.Tenv()`. Use function `std_err()` if the standard error for the tensor coefficient from `TRR.fit()` is desired.
- Improved the computation efficiency by adopting cholesky decomposition in matrix inversion and the calculation of the square root of a matrix. 
- Add a real data set `EEG`. Refer to R help documentation for more details
- Add a new feature to `predict.Tenv()`: if the argument `newdata` is missing, the fitted values from the fitted model is returned.

## Minor changes
- Change arguments `Xn` and  `Yn` in all functions to `x` and `y` in accordance with other popular functions, e.g., `lm()`, `glm()`, etc.
- `plot.Tenv()`: Change the name of argument `thrd` to `level`.
- `kroncov`: Data `Tn` is centered before the estimation. 

---

# TRES 1.1.0

## Small changes
- Add some references in Description field in file DESCRIPTION
- Fix the document of `summary.Tenv`.
- Remove the parameter `ask` in plot.Tenv. 
- Add envelope basis `Gamma` list into `bat` and `square` datasets.
- Change the names of some parameters, like `bic_max` to `maxdim` in `TensEnv_dim`, `max_iter` to `maxiter` in `manifold1D`, `epsilon` to `tol` in `ECD`, `G_ini` to `Gamma_init` in `manifoldFG`, `G_hat` to `Gamma` in `manifoldFG`, `Yhat` to `pred` in `PMSE`.
- Rename `TRR`, `TPR` as `TRR.fit`, `TPR,fit`.

## S3 methods
- `print.Tenv`: Prints the call, coefficients from `TPR` and `TRR`, make the output more concise.
- `print.summary.Tenv`: Print call, dimensions of X, Y, sample size, mse, the coefficient and p_value. (invoked implicitly when there is no assignment of `summary.Tenv`).

---

# TRES 1.0.0

## Major changes
- Switch the arguments `Xn` and `Yn` in functions: `TensEnv_dim`, `TensPLS_cv2d3d`, `Tenv`, `Tenv_Pval`, `TPR` and `TRR`.
- Removed package `mvtnorm` in `DESCRIPTION`, it is not used in the package.
- Move all the packages in `Depends` in `DESCRIPTION` to `Imports` except for `ManifoldOptim`.
- Add examples to some functions: `PMSE`, `TensEnv_dim`, `TensPLS_cv2d3d`, `Tenv_Pval`.
- Add simulated data "bat" and "square" with which users can quickly verify the functions in the package.
```
# Usage
> data("bat")
> data("square")
```
- `TPR`, `TRR`: 
  - Change the output names of some functions. E.g., `Bhat` in the object returned from `TPR` and `TRR` is renamed as `coefficients`.
  - Append `fitted.values` and `residuals` into the output.
  - More robust to the input type: `Yn` can be vector or matrix and `Xn` can be matrix, array or tensor for `TPR`, `Xn` can be vector or
  matrix and `Yn` can be matrix, array or tensor for `TRR`.

R/bat.R
#' Bat simulated data
#'
#' Synthetic data generated from tensor response regression (TRR) model. Each response observation is a two-dimensional image, and each binary predictor observation takes values 0 and 1, representing two groups.
#'
#' @docType data
#'
#' @usage data("bat")
#'
#' @format A list consisting of four components:
#' \describe{
#'  \item{x}{A \eqn{1 \times 20} matrix, each entry takes values 0 and 1, representing two groups.}
#'  \item{y}{A \eqn{64\times 64\times 20} tensor, each matrix \code{y@data[,,i]} represents an image.}
#'  \item{coeffiicients}{A \eqn{64\times 64 \times 1} tensor with the bat pattern.}
#'  \item{Gamma}{A list consisting of two \eqn{64 \times 14} envelope basis.}
#' }
#'
#' @details The dataset is generated from the tensor response regression (TRR) model:
#' \deqn{Y_i = B X_i + \epsilon_i,  i = 1,\ldots, n,}
#' where \eqn{n=20} and the regression coefficient \eqn{B \in R^{64\times 64}} is a given image with rank 14, representing the mean difference of the response \eqn{Y} between two groups. To make the model conform to the envelope structure, we construct the envelope basis \eqn{\Gamma_k} and the covariance matrices \eqn{\Sigma_k, k=1,2}, of error term as following. With the singular value decomposition of \eqn{B}, namely \eqn{B = \Gamma_1 \Lambda \Gamma_2^T}, we choose the envelope basis as \eqn{\Gamma_k \in R^{64\times 14}, k=1,2}. Then the envelope dimensions are \eqn{u_1 =  u_2 = 14}. We generate another two matrices \eqn{\Omega_k \in R^{14\times 14} = A_k A_k^T} and  \eqn{\Omega_{0k} \in R^{50\times 50} =  A_{0k}A_{0k}^T}, where \eqn{A_k \in R^{14\times 14}} and \eqn{A_{0k} \in R^{50\times 50}} are randomly generated from Uniform(0,1) elementwise. Then we set the covariance matrices \eqn{\Sigma_k = \Gamma_k\Omega_k \Gamma_k^T + \Gamma_{0k}\Omega_{0k} \Gamma_{0k}^T}, followed by normalization with their Frobenius norms. We set the first 10 predictors \eqn{X_i, i=1,\ldots, 10,} as 1 and the rest as 0. The error term is then generated from two-way tensor (matrix) normal distribution \eqn{TN( 0; \Sigma_1, \Sigma_2)}.
#'
#' @keywords datasets
#' @examples
#' ## Fit bat dataset with the tensor response regression model
#' data("bat")
#' x <- bat$x
#' y <- bat$y
#' # Model fitting with ordinary least square.
#' fit_std <- TRR.fit(x, y, method="standard")
#' # Draw the coefficient and p-value plots
#' plot(fit_std)
#'
#' @references  Li, L. and Zhang, X., 2017. Parsimonious tensor response regression. Journal of the American Statistical Association, 112(519), pp.1131-1146.
"bat"

